### ISSUE TITLE
Residuals for RollingOLS

### ISSUE THREAD

**Issue by matthewgilbert** [statsmodels/statsmodels#7521]
I'm wondering if it is possible to get residuals from the fit object of a `statsmodels.regression.rolling.RollingOLS` model? Something akin to `statsmodels.regression.linear_model.RegressionResultsWrapper.resid`, `statsmodels.regression.recursive_ls.RecursiveLSResultsWrapper.resid` or `statsmodels.regression.recursive_ls.RecursiveLSResultsWrapper.resid_recursive`?

(As an aside, @ChadFulton is there a DOI associated with Harvey 1989 that you mention [here](https://www.statsmodels.org/stable/generated/statsmodels.regression.recursive_ls.RecursiveLSResults.resid_recursive.html?highlight=Harvey#statsmodels.regression.recursive_ls.RecursiveLSResults.resid_recursive)? As someone unfamiliar with the literature googling Harvey 1989 seems to return a reference to [The condition of postmodernity by David Harvey](https://g.co/kgs/AQdFhZ) which I don't think is the relevant text?)

My impression is that calculating the residuals from the rolling regression would be fairly straightforward? I.e. some generalized version of the following

```python
import numpy
import pandas

from statsmodels.regression.rolling import RollingOLS

n = 1000
x = numpy.random.randn(n, 2)
beta = [2, 1]
y = (beta * x).sum(axis=1) + numpy.random.randn(n)

def rolling_residuals(rolling_parameters, X, y):
    y_hat = X * rolling_parameters
    return y - y_hat

rolling_ols_fit = RollingOLS(endog=y, exog=x, window=250, min_nobs=10).fit()
rolling_residuals(rolling_ols_fit.params, x, y)
```

Is there something I am overlooking for why calculating residuals in this way should be avoided?

**Comment by matthewgilbert** [comment]
Re DOI, nevermind some more searching turned up the relevant text, which I believe is

_Harvey, A. C. (1989). Forecasting, Structural Time Series Models and the
Kalman Filter. Cambridge: Cambridge University Press._

Apologies.

**Comment by ChadFulton** [comment]
I have not used rolling OLS extensively for anything myself, but I suppose that the appropriate residuals depend on what you plan to do with them.

For example, in your code you are computing residuals `e_i` as `e_i = y_i - E[y_i | y_{i-1}, y_{i-2}, ..., y-{i-w}]`, where the window size is `w`. This is computing the residual for each datapoint `i` using the regression based on observation `i` and the previous `w-1` observations.

But there is no particular reason you couldn't compute them as, for example, the datapoint `i` and the *subsequent* `w-1` observations, `e_i = y_i - E[y_i | y_{i+1}, y_{i+2}, ..., y-{i+w}]`.

**Comment by ChadFulton** [comment]
> Re DOI, apologies. some more searching turned up the relevant text
> 
> Harvey, A. (1990). Forecasting, Structural Time Series Models and the Kalman Filter. Cambridge: Cambridge University Press. doi:10.1017/CBO9781107049994
> 
> As an aside, it seems like most texts, e.g. Durbin & Koopman, reference what I believe is the original edition from 1989 but the only reference I could find online was to the reprint in 1990.

Thanks, this is a good point, and in any case we should provide the reference in the file where we cite it.

**Comment by bashtage** [comment]
I implemented `RollingOLS`. I wasn't sure what `residuals` should look like since there are at least 2 different definitions.  1. A set of T-window+1 set of residuals for each regression run, or two a single set of rolling residuals that would only include the last observation in the residual, e.g., if using observations [12, 13, 14, 15, 16], then only y[16] - x[16]@beta[16].

I think if you use pandas the second type are very easy to get

```python
import pandas as pd
import statsmodels.api as sm
import numpy as np
from statsmodels.regression.rolling import RollingOLS

index = pd.date_range("2000-1-1",periods=250, freq="M")
y = pd.Series(np.random.standard_normal(250),index=index)
x = sm.add_constant(pd.DataFrame(np.random.standard_normal((250,3)),index=index))
x = sm.add_constant(pd.DataFrame(np.random.standard_normal((250,3)),index=index))

res = RollingOLS(y,x, window=60).fit()
rolling_resid = y - (res.params * x).sum(1)
```

**Comment by matthewgilbert** [comment]
Thanks @bashtage for clarifying that the ambiguity around what `.resid` would refer to in a rolling model is why there is no implementation. I will close this as a result, with the understanding that if a user wants residuals based on the second approach you describe they can extract them as you demonstrate above.

**Comment by bashtage** [comment]
Code is pretty easy, and now this is documented.

**Comment by robertmartin8** [comment]
For people using `RollingOLS.from_formula`, this won't work because of `Intercept` and categorical encodings.

The workaround is to first transform your data using `patsy`. 

Given some `mod_str = "y ~ x1 + C(x2)"` and your dataframe `data`:

```python
from patsy import dmatrix

mod = rolling.RollingOLS.from_formula(mod_str, data=data, window=100).fit()
reg_data = dmatrix(mod_str.split("~")[1], data)
yhat = (mod.params * reg_data).sum(axis="columns")
rolling_resid = data["y"] - yhat
```

**Comment by bashtage** [comment]
I think that we should probably add a function that would handle the obvious cases.

**Comment by AnthonyTremblayy** [comment]
Has there been any further implementation for the second case (set of residuals for each regression run) in rollingOLS? I would need them to compute the standard deviations of the residuals in each run.

**Comment by bashtage** [comment]
Rolling residual variance estimators should be easy to compute along side other rolling quantities.  Since these would be a 1d array and well-defined, this doesn't create the issue that rolling residuals do.

**Comment by AnthonyTremblayy** [comment]
Yes, storing a 1d-array of residual standard deviations should be easy enough, but I was wondering if there was an easy way to access them via the attributes without having to fork the function.
